### ISSUE TITLE
[Call for testing] Keyboard enhancement for todo navigation

### ISSUE THREAD

**Issue by nutnutwin** [ransome1/sleek#140]
Hi, Thank you for your work
This one is the best app I found for todo.txt

**Did you check if a similar feature request has already been reported?**
Yes, sorry if there are any I missed

**Is your feature request related to a problem? Please describe.**
Hi, I am totally new to this app, and I am making a transition from propriety app like TickTick/Todoist
I wonder if it is possible to Navi the task list using keyboard arrow? (or anything else)
And I found no shortcut for "Toggle Completion"? 
	https://github.com/ransome1/sleek/wiki/Keyboard-shortcuts

**Describe the solution you'd like**
- Use Up/Down to move to Previous/Next task
- Use Left to Focus on the CheckBox(Enter to complete), Right to Focus on the Task(Enter to Edit)
- Use Some Shortcut(In my case Ctrl+., but any is fine) to Toggle Completion
- On Completion, Move Focus to the Next Task, preferrably. 

**Describe alternatives you've considered**
I am currently using AutoHotkey on Windows as a WorkAround
- Shift+Space ➤ Focus on Find, Tab 4 Times ➤ Focus on the First Task CheckBox
- Down ➤ Tab Twice ➤ Focus on Next Task
- Up ➤ Shift+Tab Twice ➤ Focus on Previous Task

Would be so great if you can make this, I really appreciate your work. 

Thank you.

**Comment by ransome1** [comment]
@nutnutwin, I added it to the 1.x. roadmap. But I can't say for sure when this is going to be worked on yet. As soon as there is a first testing build, I'll let you know.

**Comment by nutnutwin** [comment]
Excellent! Thank you for considering this.

**Comment by Treeoctopus** [comment]
Oh yes, yes, this would be great. I can't seem to get my head around how the keyboard up-down arrows just moves the window as if it's a web browser. For me personally my brain is entirely item-by item, and the only way it works for keyboard access now is to tab your way past both checkboxes  and line items. 

I quite like the VIM-inspired keyboard style of https://mjdescy.github.io/TodoTxtMac/, but it has nothing on this app for style,, filtering, and readability.

**Comment by ransome1** [comment]
@nutnutwin @Treeoctopus couldn't be be even more strict and efficient for the arrow left and right arrow behaviour?

- Imagine arrow left (if todo item is focused) would set complete/incomplete with no further click (no focus on checkbox and tab on Enter would be needed) and if arrow right would open the context menu. 
- Tab on Enter while a todo item / row is focused would open the todo.

How would you rate this behaviour from a usability perspective?

**Comment by ransome1** [comment]
@nutnutwin @Treeoctopus you can check out a first prototype on how the handling with arrow keys can look and feel like: https://github.com/ransome1/sleek/releases/tag/v1.1.6-rc.1

- `Arrow down` and `up` let the focus jump from one todo to the next
- `Enter` on a focused todo will open it
- `Arrow right` will open the context menu and within the menu you can continue using `arrow up` and `down` and `escape` to close it.
- `Arrow left` will set the todo complete or if it is already complete set it incomplete again

I will work to further improve this further but would appreciate your early feedback on this.

**Comment by 9z0** [comment]
Up, down and right are fine.

Left to toggle completion is a big issue. I was trying to move left in the search bar and marked a few tasks as complete by accident! It was panic-inducing but I was able to restore the previous state.
Ctrl+something is infinitely better.

**Comment by ransome1** [comment]
> able

Please do not use your actual todo data with early release candidates of sleek, I don't want to be responsible for any data loss ;) In general this of course could be fixed but disabling Arrow left, when a text field is focused. But I think it's just not a usability users would expect. 

Therefor I too think we should move the completion event to another shortcut. What about just `x` if the row is focused? That would be congruent with todotxt.net (https://github.com/benrhughes/todotxt.net#features)

**Comment by Treeoctopus** [comment]
Wow, great, I'm really glad to see the replies on this thread. I'll give the prototype a look and report back soon!

**Comment by Treeoctopus** [comment]
Here's some quick feedback. This is GREAT work, thank you muchly!

? Not sure why links/emails are shortened -- seems to serve no function. Sometimes I share multiple URL links via an Android app from the same dom	ain, these would be indistinguishable. Why not an option to do text wrap to 2, 3, 4 lines-- see screencap for both list entries and edit box?

* Could the highlighted entry have a bit more contrast against background (both light+dark mode)

* Upon program launch, an alternative way that might be a bit more useful is to have the combo "search & compose" box at top open on launch.

* I like the compose option as a fallback for the search in the top box, but why not just allow "enter" to register a new list item if no search function is called? In a related observation, could a bit more visual contrast on the add-to-todo button please be added.

* Wierdly, scrolling with two fingers on my mac scrolls in a loop, i.e. scrolling to the bottom goes right back to the top. An unexpected result.

* Pressing right on the arrow to get the popup menu: "edit" is by far more likely to be selected than "use as template", so can it be preselected for a quick press on enter. And could the box be justified left since our eyes will be already focused on the start of the line where we were reading-- note in the screenshot that the box is waaaayyy to the right.

* Another wierd interface, when I've selected an item and press the arrows to scroll up and down, in a long list it appears that the speed of the viewport and the cursor are mismatched. Eventually as a scroll in one direction the highlighted item is lost and while it still seems to be moving the cursor, I can't see it on the screen anymore.

* If you highlight an item in the middle of the list by keying down to it. Then click on an item up high in the list, the upper item is opened in the edit box. When you escape out of the box, the middle-list item is still highlighted, whereas we'd expect the recently clicked upper item to be.

* If you use the search box to find a string, you see the list of matches below. Expected behavior would be to have the first item in the list highlighted with a single tab press. Instead it cycles to "add" in the far right, then through the sidebar icons on the left. It takes 7 tab presses to get to the line item. The 6th tab takes you to the checkbox; now that there's the making of a good keyboard interface there's no need to ever have the checkbox itself tab-abble.
* 
![Screen Shot 2021-12-23 at 19 38 55](https://user-images.githubusercontent.com/14799233/147215047-483494e6-9ece-423f-8d5e-a7405e74bfb4.jpg)

![Screen Shot 2021-12-23 at 19 46 06](https://user-images.githubusercontent.com/14799233/147215008-c8c0056f-dd30-47c7-b279-22facc6c19a2.jpg)

**Comment by Treeoctopus** [comment]
And one more request, could we please have Ctrl+/- to increase/decrease font, and Ctrl-F to jump to the search box? These are familiar keyboard commands for Mac.

Thanks!

**Comment by Saturo78** [comment]
I can't scroll to the end of the list because the focus always jumps back to the beginning of the list just before.
Is it possible to change the color of the focused line in order to give it more contrast?

I would like to see the active line move down until it reaches the bottom of the screen and stay there. Similar to the way it is in Excel, for example.
Currently all entries move upwards and it is very difficult to focus on the active line.

**Comment by ransome1** [comment]
@Treeoctopus thanks a lot for your contructive feedback. I've been working on another feature mostly but gave this some more improvement as well in the latest pre-release: https://github.com/ransome1/sleek/releases/tag/v1.1.6-rc.2

> Here's some quick feedback. This is GREAT work, thank you muchly!
> 
> ? Not sure why links/emails are shortened -- seems to serve no function. Sometimes I share multiple URL links via an Android app from the same dom ain, these would be indistinguishable. Why not an option to do text wrap to 2, 3, 4 lines-- see screencap for both list entries and edit box?

The links are shortened in the todo list only because I think overlong links look messy and don't add a real value to the list. You do see the link in full once the modal is opened or you click on the icon. I like it like this but don't have an issue with introducing an option to disable shortening in the next releases. 

As for the wrapping idea I got to be honest with you, I'm not sure if I understand this correctly. Todos can actually wrap into several lines and in the modal you can switch from a one line input field into a multiple row textarea.

> 
> * Could the highlighted entry have a bit more contrast against background (both light+dark mode)

I haven't managed to look into this yet as this didn't occur to me. But I will give it some thoughts in the next pre release.

> * Upon program launch, an alternative way that might be a bit more useful is to have the combo "search & compose" box at top open on launch.

Do you have an example for this, I don't really understand what you mean. Is this related to this feature request?

> * I like the compose option as a fallback for the search in the top box, but why not just allow "enter" to register a new list item if no search function is called? In a related observation, could a bit more visual contrast on the add-to-todo button please be added.

Imo people tend to press enter after entering the search term because they expect the search query only to be sent on enter. In this case it would be confusing that new todos are being created. But anyway this is not related to this feature request, as far as I understand.

> * Wierdly, scrolling with two fingers on my mac scrolls in a loop, i.e. scrolling to the bottom goes right back to the top. An unexpected result.

If this happened it was definitely a bug. Please check this next pre release, I assume it's gone.

> * Pressing right on the arrow to get the popup menu: "edit" is by far more likely to be selected than "use as template", so can it be preselected for a quick press on enter. And could the box be justified left since our eyes will be already focused on the start of the line where we were reading-- note in the screenshot that the box is waaaayyy to the right.

If a todo is highlighted you can press enter to open it, actually there is no need to trigger the context menu by keyboard input and then select the edit entry. I'm actually thinking about removing edit from the list. 

> * Another wierd interface, when I've selected an item and press the arrows to scroll up and down, in a long list it appears that the speed of the viewport and the cursor are mismatched. Eventually as a scroll in one direction the highlighted item is lost and while it still seems to be moving the cursor, I can't see it on the screen anymore.

I think I've seen this behavior at some point as well (if I understand correctly). Can you please check if this still happens?

> * If you highlight an item in the middle of the list by keying down to it. Then click on an item up high in the list, the upper item is opened in the edit box. When you escape out of the box, the middle-list item is still highlighted, whereas we'd expect the recently clicked upper item to be.

This should not happen in the latest release any longer.

> * If you use the search box to find a string, you see the list of matches below. Expected behavior would be to have the first item in the list highlighted with a single tab press. Instead it cycles to "add" in the far right, then through the sidebar icons on the left. It takes 7 tab presses to get to the line item. The 6th tab takes you to the checkbox; now that there's the making of a good keyboard interface there's no need to ever have the checkbox itself tab-abble.

This should be improved by now. What strikes me is that the checkbox still used to be tab-abble. I never witnessed this but if this still occurs please let me know and I'll dig into it.

**Comment by ransome1** [comment]
> And one more request, could we please have Ctrl+/- to increase/decrease font, and Ctrl-F to jump to the search box? These are familiar keyboard commands for Mac.
> 
> Thanks!

The keyboard shortcuts are adopted from the todotxt.net application. Before we start changing any shortcuts we need a new and thought through concept. I think there is another feature request that deals with the general enhancement on keyboard shortcuts in sleek.

**Comment by ransome1** [comment]
> Up and down (and even right) are fine.
> 
> Left to toggle completion is a big issue. I was trying to move left in the search bar and marked a few tasks as complete by accident! It was panic-inducing but I was able to restore the previous state. Ctrl+something is infinitely better.

In the latest pre release you can press x while the todo is focused and it will mark as complete.

**Comment by ransome1** [comment]
Due to no further activity I'm closing here to clean up the backlog. If anything new pops up, please open a new issue.
